kopium do not generate struct with nesting additionalProperties

This problem could be reproduced with crd cephclusters.ceph.rook.io: https://github.com/rook/rook/blob/v1.13.5/deploy/examples/crds.yaml#L894-L912

kopium generate something like:
```rust
#[derive(CustomResource, Serialize, Deserialize, Clone, Debug, JsonSchema)]
#[kube(group = "ceph.rook.io", version = "v1", kind = "CephCluster", plural = "cephclusters")]
#[kube(namespaced)]
#[kube(status = "CephClusterStatus")]
pub struct CephClusterSpec {
    #[serde(default, skip_serializing_if = "Option::is_none")]
    pub annotations: Option<BTreeMap<String, CephClusterAnnotations>>,
    #[serde(default, skip_serializing_if = "Option::is_none", rename = "cephConfig")]
    pub ceph_config: Option<BTreeMap<String, CephClusterCephConfig>>,
    #[serde(default, skip_serializing_if = "Option::is_none", rename = "cephVersion")]
    pub ceph_version: Option<CephClusterCephVersion>,
    #[serde(default, skip_serializing_if = "Option::is_none", rename = "cleanupPolicy")]
    pub cleanup_policy: Option<CephClusterCleanupPolicy>,
    #[serde(default, skip_serializing_if = "Option::is_none", rename = "continueUpgradeAfterChecksEvenIfNotHealthy")]
    pub continue_upgrade_after_checks_even_if_not_healthy: Option<bool>,
    #[serde(default, skip_serializing_if = "Option::is_none", rename = "crashCollector")]
    pub crash_collector: Option<CephClusterCrashCollector>,
    #[serde(default, skip_serializing_if = "Option::is_none")]
    pub csi: Option<CephClusterCsi>,
    #[serde(default, skip_serializing_if = "Option::is_none")]
    pub dashboard: Option<CephClusterDashboard>,
    #[serde(default, skip_serializing_if = "Option::is_none", rename = "dataDirHostPath")]
    pub data_dir_host_path: Option<String>,
    #[serde(default, skip_serializing_if = "Option::is_none", rename = "disruptionManagement")]
    pub disruption_management: Option<CephClusterDisruptionManagement>,
    #[serde(default, skip_serializing_if = "Option::is_none")]
    pub external: Option<CephClusterExternal>,
    #[serde(default, skip_serializing_if = "Option::is_none", rename = "healthCheck")]
    pub health_check: Option<CephClusterHealthCheck>,
    #[serde(default, skip_serializing_if = "Option::is_none")]
    pub labels: Option<BTreeMap<String, CephClusterLabels>>,
    #[serde(default, skip_serializing_if = "Option::is_none", rename = "logCollector")]
    pub log_collector: Option<CephClusterLogCollector>,
    #[serde(default, skip_serializing_if = "Option::is_none")]
    pub mgr: Option<CephClusterMgr>,
    #[serde(default, skip_serializing_if = "Option::is_none")]
    pub mon: Option<CephClusterMon>,
    #[serde(default, skip_serializing_if = "Option::is_none")]
    pub monitoring: Option<CephClusterMonitoring>,
    #[serde(default, skip_serializing_if = "Option::is_none")]
    pub network: Option<CephClusterNetwork>,
    #[serde(default, skip_serializing_if = "Option::is_none")]
    pub placement: Option<BTreeMap<String, CephClusterPlacement>>,
    #[serde(default, skip_serializing_if = "Option::is_none", rename = "priorityClassNames")]
    pub priority_class_names: Option<BTreeMap<String, String>>,
    #[serde(default, skip_serializing_if = "Option::is_none", rename = "removeOSDsIfOutAndSafeToRemove")]
    pub remove_os_ds_if_out_and_safe_to_remove: Option<bool>,
    #[serde(default, skip_serializing_if = "Option::is_none")]
    pub resources: Option<BTreeMap<String, CephClusterResources>>,
    #[serde(default, skip_serializing_if = "Option::is_none")]
    pub security: Option<CephClusterSecurity>,
    #[serde(default, skip_serializing_if = "Option::is_none", rename = "skipUpgradeChecks")]
    pub skip_upgrade_checks: Option<bool>,
    #[serde(default, skip_serializing_if = "Option::is_none")]
    pub storage: Option<CephClusterStorage>,
    #[serde(default, skip_serializing_if = "Option::is_none", rename = "waitTimeoutForHealthyOSDInMinutes")]
    pub wait_timeout_for_healthy_osd_in_minutes: Option<i64>,
}
```


and failed to compile:
```
error[E0412]: cannot find type `CephClusterAnnotations` in this scope
   --> src/crds/cephclusters_ceph_rook_io.rs:17:46
    |
17  |     pub annotations: Option<BTreeMap<String, CephClusterAnnotations>>,
    |                                              ^^^^^^^^^^^^^^^^^^^^^^
...
846 | pub struct CephClusterMonitoring {
    | -------------------------------- similarly named struct `CephClusterMonitoring` defined here
    |
help: a struct with a similar name exists
    |
17  |     pub annotations: Option<BTreeMap<String, CephClusterMonitoring>>,
    |                                              ~~~~~~~~~~~~~~~~~~~~~
help: you might be missing a type parameter
    |
15  | pub struct CephClusterSpec<CephClusterAnnotations> {
    |                           ++++++++++++++++++++++++

error[E0412]: cannot find type `CephClusterCephConfig` in this scope
  --> src/crds/cephclusters_ceph_rook_io.rs:19:46
   |
19 |     pub ceph_config: Option<BTreeMap<String, CephClusterCephConfig>>,
   |                                              ^^^^^^^^^^^^^^^^^^^^^
...
71 | pub struct CephClusterCephVersion {
   | --------------------------------- similarly named struct `CephClusterCephVersion` defined here
   |
help: a struct with a similar name exists
   |
19 |     pub ceph_config: Option<BTreeMap<String, CephClusterCephVersion>>,
   |                                              ~~~~~~~~~~~~~~~~~~~~~~
help: you might be missing a type parameter
   |
15 | pub struct CephClusterSpec<CephClusterCephConfig> {
   |                           +++++++++++++++++++++++

error[E0412]: cannot find type `CephClusterLabels` in this scope
  --> src/crds/cephclusters_ceph_rook_io.rs:41:41
   |
41 |     pub labels: Option<BTreeMap<String, CephClusterLabels>>,
   |                                         ^^^^^^^^^^^^^^^^^ not found in this scope
   |
help: you might be missing a type parameter
   |
15 | pub struct CephClusterSpec<CephClusterLabels> {
   |                           +++++++++++++++++++

error[E0412]: cannot find type `CephClusterAnnotations` in this scope
   --> src/crds/cephclusters_ceph_rook_io.rs:17:46
    |
17  |     pub annotations: Option<BTreeMap<String, CephClusterAnnotations>>,
    |                                              ^^^^^^^^^^^^^^^^^^^^^^ help: a struct with a similar name exists: `CephClusterMonitoring`
...
846 | pub struct CephClusterMonitoring {
    | -------------------------------- similarly named struct `CephClusterMonitoring` defined here

error[E0412]: cannot find type `CephClusterCephConfig` in this scope
  --> src/crds/cephclusters_ceph_rook_io.rs:19:46
   |
19 |     pub ceph_config: Option<BTreeMap<String, CephClusterCephConfig>>,
   |                                              ^^^^^^^^^^^^^^^^^^^^^ help: a struct with a similar name exists: `CephClusterCephVersion`
...
71 | pub struct CephClusterCephVersion {
   | --------------------------------- similarly named struct `CephClusterCephVersion` defined here

error[E0412]: cannot find type `CephClusterLabels` in this scope
  --> src/crds/cephclusters_ceph_rook_io.rs:41:41
   |
41 |     pub labels: Option<BTreeMap<String, CephClusterLabels>>,
   |                                         ^^^^^^^^^^^^^^^^^ not found in this scope

For more information about this error, try `rustc --explain E0412`.
error: could not compile `operator` (bin "operator") due to 6 previous errors
```

--- a/src/analyzer.rs
+++ b/src/analyzer.rs
@@ -245,23 +245,7 @@ fn extract_container(
     for (key, value) in props {
         let value_type = value.type_.clone().unwrap_or_default();
         let rust_type = match value_type.as_ref() {
-            "object" => {
-                let mut dict_key = None;
-                if let Some(additional) = &value.additional_properties {
-                    dict_key = resolve_additional_properties(additional, stack, key)?;
-                } else if value.properties.is_none()
-                    && value.x_kubernetes_preserve_unknown_fields.unwrap_or(false)
-                {
-                    dict_key = Some("serde_json::Value".into());
-                }
-                if let Some(dict) = dict_key {
-                    format!("{}<String, {}>", cfg.map.name(), dict)
-                } else if !cfg.no_object_reference && is_object_ref(value) {
-                    "ObjectReference".into()
-                } else {
-                    format!("{}{}", stack, key.to_upper_camel_case())
-                }
-            }
+            "object" => extract_object_type(value, stack, key, cfg)?,
             "string" => {
                 if let Some(_en) = &value.enum_ {
                     trace!("got enum string: {}", serde_json::to_string(&schema).unwrap());
@@ -363,10 +347,7 @@ fn resolve_additional_properties(
         // authoratative, but more detailed sources than crd validation docs below are welcome
         // https://kubernetes.io/docs/tasks/extend-kubernetes/custom-resources/custom-resource-definitions/#validation
         "array" => Some(array_recurse_for_type(s, stack, key, 1, cfg)?.0),
-        "object" => {
-            // cluster test with `failureDomains` uses this spec format
-            Some(format!("{}{}", stack, key.to_upper_camel_case()))
-        }
+        "object" => Some(extract_object_type(s, stack, key, cfg)?),
         "" => {
             if s.x_kubernetes_int_or_string.is_some() {
                 Some("IntOrString".into())
@@ -403,19 +384,7 @@ fn array_recurse_for_type(
                 let inner_array_type = s.type_.clone().unwrap_or_default();
                 match inner_array_type.as_ref() {
                     "object" => {
-                        // Same logic as in `extract_container` to simplify types to maps.
-                        let mut dict_value = None;
-                        if let Some(additional) = &s.additional_properties {
-                            dict_value = resolve_additional_properties(additional, stack, key, cfg)?;
-                        }
-
-                        let vec_value = if let Some(dict_value) = dict_value {
-                            let map_type = cfg.map.name();
-                            format!("{map_type}<String, {dict_value}>")
-                        } else {
-                            let structsuffix = key.to_upper_camel_case();
-                            format!("{stack}{structsuffix}")
-                        };
+                        let vec_value = extract_object_type(s, stack, key, cfg)?;
 
                         Ok((format!("Vec<{}>", vec_value), level))
                     }
@@ -501,6 +470,27 @@ fn is_object_ref(value: &JSONSchemaProps) -> bool {
         }
     }
     false
+}
+
+fn extract_object_type(
+    value: &JSONSchemaProps,
+    stack: &str,
+    key: &str,
+    cfg: &Config,
+) -> Result<String, anyhow::Error> {
+    let mut dict_key = None;
+    if let Some(additional) = &value.additional_properties {
+        dict_key = resolve_additional_properties(additional, stack, key, cfg)?;
+    } else if value.properties.is_none() && value.x_kubernetes_preserve_unknown_fields.unwrap_or(false) {
+        dict_key = Some("serde_json::Value".into());
+    }
+    Ok(if let Some(dict) = dict_key {
+        format!("{}<String, {}>", cfg.map.name(), dict)
+    } else if !cfg.no_object_reference && is_object_ref(value) {
+        "ObjectReference".into()
+    } else {
+        format!("{}{}", stack, key.to_upper_camel_case())
+    })
 }
 
 fn extract_date_type(value: &JSONSchemaProps) -> Result<String> {
@@ -623,6 +613,40 @@ mod test {
         assert_eq!(other.members[1].type_, "String");
         assert_eq!(other.members[2].name, "status");
         assert_eq!(other.members[2].type_, "String");
+    }
+
+    #[test]
+    fn map_of_map() {
+        init();
+        // as found in cnpg-cluster
+        let schema_str = r#"
+        description: Instances topology.
+        properties:
+          instances:
+            additionalProperties:
+              additionalProperties:
+                type: string
+              description: PodTopologyLabels represent the topology of a Pod.
+                map[labelName]labelValue
+              type: object
+            description: Instances contains the pod topology of the instances
+            type: object
+        type: object
+"#;
+        let schema: JSONSchemaProps = serde_yaml::from_str(schema_str).unwrap();
+        //println!("schema: {}", serde_json::to_string_pretty(&schema).unwrap());
+
+        let structs = analyze(schema, "ClusterStatusTopology", Cfg::default())
+            .unwrap()
+            .0;
+        //println!("{:?}", structs);
+        let root = &structs[0];
+        assert_eq!(root.name, "ClusterStatusTopology");
+        assert_eq!(root.level, 0);
+        // should have a member with a key to the map:
+        let map = &root.members[0];
+        assert_eq!(map.name, "instances");
+        assert_eq!(map.type_, "Option<BTreeMap<String, BTreeMap<String, String>>>");
     }
 
     #[test]

